tokio/src/io/driver/scheduled_io.rs

```diff
@@ -57,51 +134,330 @@ impl ScheduledIo {
+                            *state = State::Done;
+                            return Poll::Ready(ReadyEvent { readiness, tick });
+                        }
+
+                        // Not ready even after locked, insert into list...
+
+                        // Safety: called while locked
+                        unsafe {
+                            (*waiter.get()).waker = Some(cx.waker().clone());
+                        }
+
+                        // Insert the waiter into the linked list
+                        //
+                        // safety: pointers from `UnsafeCell` are never null.
+                        waiters
+                            .list
+                            .push_front(unsafe { NonNull::new_unchecked(waiter.get()) });
+                        *state = State::Waiting;
+                    }
+                    State::Waiting => {
+                        // Currently in the "Waiting" state, implying the caller has
+                        // a waiter stored in the waiter list (guarded by
+                        // `notify.waiters`). In order to access the waker fields,
+                        // we must hold the lock.
+
+                        let waiters = scheduled_io.waiters.lock().unwrap();
+
+                        // Safety: called while locked
+                        let w = unsafe { &mut *waiter.get() };
+
+                        if w.is_ready {
+                            // Our waker has been notified.
+                            *state = State::Done;
+                        } else {
+                            // Update the waker, if necessary.
+                            if !w.waker.as_ref().unwrap().will_wake(cx.waker()) {
+                                w.waker = Some(cx.waker().clone());
+                            }
+
+                            return Poll::Pending;
+                        }
+
+                        // Explicit drop of the lock to indicate the scope that the
+                        // lock is held. Because holding the lock is required to
+                        // ensure safe access to fields not held within the lock, it
+                        // is helpful to visualize the scope of the critical
+                        // section.
+                        drop(waiters);
+                    }
+                    State::Done => {
+                        let tick = TICK.unpack(scheduled_io.readiness.load(Acquire)) as u8;
+
+                        // Safety: State::Done means it is no longer shared
+                        let w = unsafe { &mut *waiter.get() };
+
+                        return Poll::Ready(ReadyEvent {
+                            tick,
+                            readiness: w.interest,
+                        });
+                    }
+                }
+            }
+        }
+    }
+
+    impl Drop for Readiness<'_> {
+        fn drop(&mut self) {
+            let mut waiters = self.scheduled_io.waiters.lock().unwrap();
```

Comment:
since removing the waiter from the list is necessary to maintain safety invariants, i think we should change this to use `LockResult::into_inner` to extract a mutex guard even if the lock was poisoned. if we don't ensure it's removed from the list, we'll leave the list with a potentially dangling pointer that is assumed to be valid.

Reply:
The idea of relying on `Drop` for safety is scary and probably wrong...

Also, I borrowed a lot of this from `notify`, like: https://github.com/tokio-rs/tokio/blob/e7091fde786722a5301270e6281fc3c449dcfc14/tokio/src/sync/notify.rs#L560

Reply:
> The idea of relying on Drop for safety is scary and probably wrong...

`Pin` guarantees that the memory location remains valid until it is dropped. If it is not dropped, then the memory must be leaked and remain valid forever.

Reply:
> The idea of relying on `Drop` for safety is scary and probably wrong...

We're relying on the drop impl to maintain safety regardless of whether or not we ignore mutex poisoning here, it's inherent to the intrusive list that nodes must be unlinked before they are deallocated.

I think `Notify` also needs to needs to ignore poisoning for the same reason. It might actually matter more in that case since there are probably going to be higher-level users of `Notify` that send a notification when dropped, potentially including while unwinding...

Reply:
I think this change should be done in a dedicated PR :) I tracked it here: https://github.com/tokio-rs/tokio/issues/2867

For now, `unwrap()` is fine.